sky/serve/replica_managers.py

```diff
@@ -705,6 +766,13 @@ def _refresh_process_pool(self) -> None:
                         removal_reason = 'normally'
                     logger.info(f'Replica {replica_id} removed from the '
                                 f'replica table {removal_reason}.')
+                # Don't keep failed record for version mismatch replicas,
+                # since user should fixed the error before update.
+                elif info.version != self.latest_version:
+                    serve_state.remove_replica(self._service_name, replica_id)
+                    logger.info(
+                        f'Old replica {replica_id} (version {info.version}) '
+                        f'removed from the replica table.')
```

Comment:
nit: we should combine the two case when it is possible to remove an replica and maintain an flag (say `removal_reason is None`) to trigger a removal of replicas. e.g.

```python
removal_reason = None
if case1:
    if xxx:
        removal_reason = '1'
    else:
        removal_reason = '2'
elif case2:
    removal_reason = '3'
else:
    # do sth
if removal reason is not None:
    serve_state.remove_replica(...)
```